[DOC] Adds developer guideline start (#446)
ME-ICA/tedana: CONTRIBUTING.md

Already know what you're looking for in this guide? Jump to the following sections:

> 

Reviewer: This is a very useful section, I think there are a couple of other points in the document readers might want to quickly navigate to, so I would add:
* [installing in editable mode](#Run-the-developer-setup)
* [testing and linting](#Test-your-changes)
* [updating the documentation after changes to the code](#Changes-to-documentation)

I would also link to the developer docs here with a brief description of the contents, say: 
See also our Developing Guidelines on [readthedocs][link_developing_rtd] for additional information, particularly on our monthly developer calls and writing tests for tedana.

There is significant cross-talk between these two spaces, and we look forward to hearing from you in either venue!
As a reminder, we expect all contributions to `tedana` to adhere to our [code of conduct][link_coc].

> 

Author: Yes, I must have somehow gotten distracted and not pasted it in. Addressed in aa41ffe, with a slight modification since the agenda is often last-minute.

### 4. Make the changes you've discussed

Try to keep the changes focused to the issue.
We've found that working on a [new branch][link_branches] for each issue makes it easier to keep your changes targeted.
Using a new branch allows you to follow the standard GitHub workflow when making changes.
[This guide][link_gitworkflow] provides a useful overview for this workflow.

> Before making a new branch, make sure your master is up to date.

Author: ```suggestion
Before making a new branch, make sure your master is up to date with the following commands:
```

Author: @Islast how does that sound?

Reviewer: perfect :+1:

> ```

Reviewer: maybe it's worth adding a sentence to the effect of "you can do this by running the following in your command line" in front of this code snippet

### 4. Make the changes you've discussed

Try to keep the changes focused to the issue.  
We've found that working on a [new branch][link_branches] for each issue makes it easier to keep your changes targeted.
Using a new branch allows you to follow the standard "fork/branch/commit/pull-request/merge" GitHub 
workflow when making changes.
[This guide][link_gitworkflow] provides a useful overview for this workflow.
Before making a new branch, make sure your master is up to date.

```
git checkout master
git fetch upstream master
git merge upstream/master
```

Then, make your new branch.

```
git checkout -b MYBRANCH
```

> As you're making changes, make sure your branch is kept up to date with 

Author: These usually end up getting erased as we squash and merge, though, right?

> As you're making changes, make sure your branch is kept up to date with 

Author: @emdupre do you mind commenting really quickly on this?

> As you're making changes, make sure your branch is kept up to date with 

Reviewer 2: I'm leaning against it only because I think it's not necessary unless there's a merge conflict !

> As you're making changes, make sure your branch is kept up to date with 

Author: Hm, I see your point, I'll reword it. For novices that could be confusing.

```
git fetch upstream master
git merge upstream/master
```

> If you know what rebasing is, please only use it for changes that haven't been pushed.

Author: The problem I've run into is that new git users will look up how to do something, and then see a StackExchange post saying to rebase or some forum post saying to do so, and then they do so without realizing the consequences. For example, the Linux kernel frequently recommends a rebase, but they're also targeting users much more git-familiar than a typical contributor in open source science projects. This has happened a few times, so I would like to do something to indicate to people that we don't recommend it unless they're sure of what they're doing.

> If you know what rebasing is, please only use it for changes that haven't been pushed.

Reviewer 2: Maybe we could say something more generally about not amending history in pull requests ? So it would cover git commit --amend, too ? We can link here: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

And point out that rewriting history can make it more difficult for others to work collaboratively with you on a new feature or bug fix ?

> If you know what rebasing is, please only use it for changes that haven't been pushed.

Author: That makes sense. I always forget that you can `--amend` because usually I do my cleanup as an interactive rebase. I probably should have dropped the link in there to begin with, it's the link I use in my tutorials for my lab. I'll go ahead and work it into that section, thanks!

> If you know what rebasing is, please only use it for changes that haven't been pushed.

Author: Do you like how it's handled in 1bd9b19?

> If you know what rebasing is, please only use it for changes that haven't been pushed.

Author: I've opted to just delete this part since we like how it's handled in the other section. I figure it's somewhat unlikely for someone to attempt a rebase at that point unless they're aware of that version of the workflow.

If you don't know what rebasing is, don't do it at all, as it is the easiest way to make your 
repository a disaster zone.
Please make sure to review the `tedana` [style conventions](#style-guide) and test your changes.

If you are new to ``git``, there are several GUI git clients that you may find helpful, such as
- [GitKraken][link_git_kraken]
- [GitHub Desktop][link_github_desktop]
- [SourceTree][link_source_tree]

> ### 5. Test your changes

Reviewer 2: One general thing -- I think the way this section is currently written risks discouraging people from opening pull requests "early and often" which is something we explicitly want.

Especially since the testing is so minimal, I'd almost rather we didn't include this section and instead wrote something about how we expect all new code to include tests, and guides for writing good tests. WDYT ?

Author: That sounds like a good alternative. I can see why the phrasing I had is discouraging, I'll amend it ahead of Thanksgiving.

Author: @emdupre do you consider this resolved now since you approved earlier?

You can run style checks by running the following:
```
flake8 $TEDANADIR/tedana
```

and unit/integration tests by running `pytest` (more details below).

> If you know a file will test your change, you can run only that test.

Author: Addressed in 267bf66

Alternatively, running all unit tests is relatively quick and should be
fairly comprehensive.
Running all `pytest` tests will be useful for pre-pushing checks.
When you open a Pull Request, CircleCI will run all tests.

```

> # All tests; final checks before pushing

Author: I'm not sure I follow, sorry, could you elaborate?

> # All tests; final checks before pushing

Reviewer 3: I think (in line with @emdupre's comment) it would be great to take each of the `pytest` commands below and simply add them beneath each of the descriptive sentences above. For example:

> If you know a file will test your change, you can run only that test with:
> 
> ```
> pytest -k my_function $TEDANADIR/tedana/tests/test_file.py
> ```

and so on.

pytest $TEDANADIR/tedana/tests
# Unit tests and linting only
pytest --skipintegration $TEDANADIR/tedana/tests
# One test file in particular
pytest $TEDANADIR/tedana/tests/test_file.py
# Test one function in a file
pytest -k my_function $TEDANADIR/tedana/tests/test_file.py
```

from within your local `tedana` repository.
The test run will indicate the number of passes and failures.

> Most often, the failures give enough information to determine the cause; if not, if not you can

Reviewer 3: ```suggestion
Most often, the failures give enough information to determine the cause; if not, you can
```

### 5. Test your changes

You can run style checks by running the following

```

> cd TEDANADIR

Reviewer 2: Can you explain that `TEDANADIR` is a file path, here ? And, can we mark it as a variable with a `$` ?

flake8 tedana
```

> and unit/integration tests by running the following:

Reviewer 2: Since you give a number of options below, can we either add more explanatory text here or break up that block to reference the different calls separately ?

```

> cd TEDANADIR/tedana/tests

Reviewer 2: I don't think you need to navigate into the directory to run pytest. It should find them just from the $TEDANADIR  with `pytest tedana`.

# All tests
pytest .
# Unit tests and linting only (saves time)
pytest --skipintegration .
# One test file in particular
pytest test_file.py
# Test one function in a file
pytest -k my_function test_file.py
```

from within your local `tedana` repository.

> The test run will indicate the number of passes and failures.

Reviewer 2: Should we explain these are run as part of the automated testing, too ?

* **[ENH]** for enhancements
* **[FIX]** for bug fixes
* **[REF]** for refactoring existing code
* **[TST]** for new or updated tests, and
* **[MAINT]** for maintenance of code

You can also combine the tags above, for example if you are updating both a test and
the documentation: **[TST, DOC]**.

Pull requests should be submitted early and often!
If your pull request is not yet ready to be merged, please use [draft PRs][link_draftpr]
This tells the development team that your pull request is a "work-in-progress",
and that you plan to continue working on it.
Note that if your pull request has no conversation or commits for 90 days,

> stale-bot will mark it stale and we may ask your permission to take over it.

Reviewer 2: ```suggestion
stale-bot will mark it stale.
We may also ask that another development team member resume work on it, if you are unable to continue to do so.
```

> Pull Request Checklist (Before Requesting Review):

Reviewer 2: Can this be a heading ?

* **[ENH]** for enhancements
* **[FIX]** for bug fixes
* **[REF]** for refactoring existing code
* **[TST]** for new or updated tests, and
* **[MAINT]** for maintenance of code

You can also combine the tags above, for example if you are updating both a test and
the documentation: **[TST, DOC]**.

Pull requests should be submitted early and often!
If your pull request is not yet ready to be merged, please use [draft PRs][link_draftpr]
This tells the development team that your pull request is a "work-in-progress",
and that you plan to continue working on it.
Note that if your pull request has no conversation or commits for 90 days,
stale-bot will mark it stale.
We may also ask that another development team member resume work on it, if you are unable to continue to do so.

> ### Pull Request Checklist (Before Requesting Review):

Reviewer 2: Again, we want to make sure we're not discouraging people from _opening_ pull requests, just setting expectations around what will get a full review.

- [ ] Check that all tests are passing ("All tests passsed")
- [ ] Make sure you have docstrings for any new functions
- [ ] Make sure that docstrings are updated for edited functions
- [ ] Make sure you note any issues that will be closed by your PR
- [ ] Take a look at the automatically generated readthedocs for your PR (Show all checks -> continuous-documentation/readthedocs -> Details)

### After Changes Are Merged
After successful merging of the pull request, remember to [keep your fork up to date][link_updateupstreamwiki] with the master `tedana` repository and to delete the branch on your fork that was used for the merged pull request.

> You may also want to update your other branches to include these changes,

Reviewer 2: I think this is redundant with the idea of merging in master, which I'd again encourage folks not to use unless there's a merge conflict.

Reviewer 3: Agreed. Since we're suggesting as a first step in the contribution process that users fetch & merge upstream/master (before creating a new branch), I think we can safely remove this section!

Author: Okay, I've just deleted those sections.

which you can do via

```
git checkout master
git fetch upstream master
git merge upstream/master
git checkout OTHERBRANCH
git merge master
```

> In the event of confusion, please ping in the Gitter for help.

Reviewer 2: Let's not recommend Gitter, here.

Author: 🤣 good point.

Author: Alternative of opening Draft PR and requesting help is proposed in 6aa2296

> We strongly recommend you *not* use the `git rebase` command when working with and contributing to the `tedana` repository.

Reviewer 2: Can we drop the mention of rebase here, too ?

Author: See other comment.

> ### Advanced Development

Reviewer 2: Unclear what this means to a naive reader. Can you include examples (e.g., `such as....`) ?

Author: I'm not sure that I have a great definition here; perhaps we could say something like, "Comprehensive Developer Guidelines" instead?

Author: I just renamed it "Comprehensive Developer Guidelines" in b7374a1

stale-bot will mark it stale.
We may also ask that another development team member resume work on it, if you are unable to continue to do so.

###Pull Request Checklist (Before Requesting Review):
- [ ] Check that all tests are passing ("All tests passsed")
- [ ] Make sure you have docstrings for any new functions
- [ ] Make sure that docstrings are updated for edited functions
- [ ] Make sure you note any issues that will be closed by your PR
- [ ] Take a look at the automatically generated readthedocs for your PR (Show all checks -> continuous-documentation/readthedocs -> Details)

### After Changes Are Merged
After successful merging of the pull request, remember to [keep your fork up to date][link_updateupstreamwiki] with the master `tedana` repository and to delete the branch on your fork that was used for the merged pull request.
You may also want to update your other branches to include these changes,
which you can do via

```
git checkout master
git fetch upstream master
git merge upstream/master
git checkout OTHERBRANCH
git merge master
```

In the event of confusion, please ping in the Gitter for help.

> BE WARY OF WELL-INTENTIONED STACK OVERFLOW USERS TELLING YOU TO REBASE.

Author: It was originally intended for use on the RTD, which is more removed from new users, but good point now that it's in CONTRIBUTING
